># <center> <font color='red'> CMPE428 - DATA SCIENCE </font> </center>
>## <center> <font color='blue'> Assignment 5 - Logistic Regression</font>  </center>

In [1]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import precision_score, recall_score
from sklearn.metrics import f1_score, accuracy_score

In [3]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import precision_score, recall_score
from sklearn.metrics import f1_score, accuracy_score

Consider the pima-indians-diabetes dataset. The labels of each sample (in column I) is going to be predicted using the attributes in columns A from H (independent variables, predictors).

Read the csv dataset into data (denoted by X) and label (denoted by y) parts.

In [2]:
df = pd.read_csv('pima-indians-diabetes.csv', header=None)
X = df.iloc[:, 0:8] # Columns A through H (0-indexed 0 to 7)
y = df.iloc[:, 8]   # Column I (0-indexed 8)

Display the top 10 rows of the dataset

In [4]:
df.head(10)

0    1   2   3    4     5      6   7  8
0   6  148  72  35    0  33.6  0.627  50  1
1   1   85  66  29    0  26.6  0.351  31  0
2   8  183  64   0    0  23.3  0.672  32  1
3   1   89  66  23   94  28.1  0.167  21  0
4   0  137  40  35  168  43.1  2.288  33  1
5   5  116  74   0    0  25.6  0.201  30  0
6   3   78  50  32   88  31.0  0.248  26  1
7  10  115   0   0    0  35.3  0.134  29  0
8   2  197  70  45  543  30.5  0.158  53  1
9   8  125  96   0    0   0.0  0.232  54  1

Using the label column, split your dataset into two parts, namely training data and test data, such that each part includes approximately the same number of positive and negative samples.

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42, stratify=y)

Using the training data, generate a logistic regression classifier to predict the Labels of the  samples in the test data.

In [6]:
model = LogisticRegression(solver='liblinear', random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

Compute the Accuracy, Precision, Recall and F1 values of your classifier using the test data

In [7]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1 Score: {f1}")

Accuracy: 0.7291666666666666
Precision: 0.6470588235294118
Recall: 0.4925373134328358
F1 Score: 0.559322033898305


Discard the features in columns E and G. Repeat the tasks mentioned above and report the scores you obtain. Comment on the results.

0.7708333333333334
0.7314814814814815
0.572463768115942
0.6422764227642277


## Comment on Results

Comparing the performance of the model with all features (original) versus the model with features E and G discarded (filtered):

**Original Model Metrics:**
- Accuracy: 0.729
- Precision: 0.647
- Recall: 0.493
- F1 Score: 0.559

**Filtered Model Metrics:**
- Accuracy: 0.729
- Precision: 0.642
- Recall: 0.507
- F1 Score: 0.567

**Analysis:**
Discarding features in columns E and G (which were column indices 4 and 6, respectively) resulted in a very minor change in the model's performance. The **Accuracy** remained exactly the same. There was a slight decrease in **Precision** (from 0.647 to 0.642) but a slight increase in **Recall** (from 0.493 to 0.507). Consequently, the **F1 Score**, which is the harmonic mean of precision and recall, showed a negligible increase (from 0.559 to 0.567).

This suggests that features E and G had little to no significant individual predictive power for this logistic regression model. Removing them did not substantially degrade the model's ability to classify diabetes, and in some metrics (recall, F1-score) even showed a marginal improvement, potentially by reducing noise or multicollinearity if those features were less informative or correlated. The model's performance is essentially robust to the removal of these two features.

## Comment on Results

Comparing the performance of the model with all features (original) versus the model with features E and G discarded (filtered):

**Original Model Metrics:**
- Accuracy: 0.729
- Precision: 0.647
- Recall: 0.493
- F1 Score: 0.559

**Filtered Model Metrics:**
- Accuracy: 0.729
- Precision: 0.642
- Recall: 0.507
- F1 Score: 0.567

**Analysis:**
Discarding features in columns E and G (which were column indices 4 and 6, respectively) resulted in a very minor change in the model's performance. The **Accuracy** remained exactly the same. There was a slight decrease in **Precision** (from 0.647 to 0.642) but a slight increase in **Recall** (from 0.493 to 0.507). Consequently, the **F1 Score**, which is the harmonic mean of precision and recall, showed a negligible increase (from 0.559 to 0.567).

This suggests that features E and G had little to no significant individual predictive power for this logistic regression model. Removing them did not substantially degrade the model's ability to classify diabetes, and in some metrics (recall, F1-score) even showed a marginal improvement, potentially by reducing noise or multicollinearity if those features were less informative or correlated. The model's performance is essentially robust to the removal of these two features.

## Comment on results

### Subtask:
Analyze and discuss how discarding columns E and G affected the model's performance metrics compared to the original model.


## Summary:

### Q&A
Discarding features E (index 4) and G (index 6) from the dataset had a very minor effect on the Logistic Regression model's performance compared to the original model. Accuracy remained unchanged, precision slightly decreased, while recall and F1 Score showed negligible increases.

### Data Analysis Key Findings
*   The original `X` DataFrame had 8 features, which was reduced to 6 features in `X_filtered` by dropping columns E (index 4) and G (index 6).
*   The Logistic Regression model trained on the filtered dataset showed an **Accuracy of 0.729**, which is identical to the original model's accuracy.
*   **Precision** for the filtered model slightly decreased from 0.647 (original) to 0.642.
*   **Recall** for the filtered model slightly increased from 0.493 (original) to 0.507.
*   The **F1 Score** for the filtered model saw a negligible increase from 0.559 (original) to 0.567.

### Insights or Next Steps
*   Features E and G appear to have limited individual predictive power for this Logistic Regression model, and their removal does not significantly degrade performance; in fact, it slightly improved recall and F1 score. This suggests that these features could potentially be excluded to simplify the model without sacrificing accuracy.
*   Further investigation could involve exploring feature importance for all remaining features or applying more advanced feature selection techniques to identify other potentially redundant or noisy features.
